lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchersyncer
+
+import (
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+const (
+	maxUpdatesToConsolidate = 1000
+)
+
+// ResourceType groups together the watch and conversion information for a
+// specific resource type.
+type ResourceType struct {
+	// The ListInterface used to perform a watch on this resource type.
+	ListInterface model.ListInterface
+
+	// An update processor used to convert the update event prior to it being sent in
+	// the Syncer.
+	UpdateProcessor SyncerUpdateProcessor
+}
+
+// SyncerUpdateProcessor is used to convert a Watch update into one or more additional
+// Syncer updates.
+type SyncerUpdateProcessor interface {
+	// Process is called to process a watch update.  The processor may convert this
+	// to zero or more updates.  The processor may use these calls to maintain a local cache
+	// if required.  It is safe for the processor to send multiple duplicate adds or deletes
+	// since the WatcherSyncer maintains it's own cache and will swallow duplicates.
+	// A KVPair with a nil value indicates a delete.  A non nil value indicates an add/modified.
+	// The processor may respond with any number of adds or deletes.
+	Process(*model.KVPair) ([]*model.KVPair, error)
+
+	// SyncStarting is called when syncer is starting a full sync for the associated resource
+	// type.  That means it is first going to list current resources and then watch for any updates.
+	// If the processor maintains a private internal cache, then the cache should be cleared at
+	// this point since the cache will be re-populated from the sync.
+	SyncStarting()
+}
+
+// New creates a new multiple Watcher-backed api.Syncer.
+func New(client api.Client, resourceTypes []ResourceType, callbacks api.SyncerCallbacks) api.Syncer {
+	rs := &watcherSyncer{
+		watcherCaches: make([]*watcherCache, len(resourceTypes)),
+		results:       make(chan interface{}, 2000),
```

Comment:
What's this magic number of 2000? Do we need a buffered channel here? I would have thought we shouldn't limit the length of this channel, or at least make it configurable.

Reply:
A buffered channel its sensible - it's possible that conversion between the v2 and v1 takes a chunk of time that could be done in parallel with the callback processing.  In which case having them buffered allows for more upfront concurrent processing.

Reply:
Just wondering if this is going to hurt us when we're resyncing large quantities of data. 

e.g. running a 5k node * 30 pod test = 150k pod events at start of day. We might just hit a resync loop where we fall behind because the wathcerCache for pods is stuck waiting for the queue to empty.. or maybe we're fine and we handle it quick enough anyway?

Reply:
The 2000 wasn't quite random....  The etcdv2 one had a buffered channel size of 20000 ... which i then misread as 2000.  :-/

Reply:
I don't think having a buffered channel is making things any worse is it?   Having a unbuffered one that blocked would certainly be worse.

What's the alternative that you might be eluding to?  Or do you think the channel buffer should be bigger *back to the 20000 that the etcdv2 one was).

Probable one for @fasaxc .

Reply:
Yeah, I think the question is just "is 2000 big enough or should we put it back to 20k?"